### PR TITLE
Data Split API

### LINKED ISSUE #30: Avoiding Loading dataset multiple times for Test / Val in Hooks and Graph Construction
Currently when running 

```
train_dg = DGraph(args.dataset, time_delta=TimeDeltaDG('r'), split='train')
val_dg = DGraph(args.dataset, time_delta=TimeDeltaDG('r'), split='val')
test_dg = DGraph(args.dataset, time_delta=TimeDeltaDG('r'), split='test')

train_data = train_dg.materialize(materialize_features=False)
val_loader = DGDataLoader(
    val_dg,
    hook=[TGBNegativeEdgeSamplerHook(neg_sampler, split_mode='val')],
    batch_size=args.bsize,
)
test_loader = DGDataLoader(
    test_dg,
    hook=[TGBNegativeEdgeSamplerHook(neg_sampler, split_mode='test')],
    batch_size=args.bsize,
)
```

The TGB dataset will be loaded four times, three times for DGraph for each split and one time for negative sampler.

### PR DESCRIPTION
**Pull request by Jacob-Chmura**
Close #30

### CODE REVIEW

File: tgm/split.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    def __post_init__(self) -> None:
+        total = self.train_ratio + self.val_ratio + self.test_ratio
+        if abs(total - 1.0) > 1e-6:
+            raise ValueError(
+                f'train_ratio + val_ratio + test_ratio must sum to 1.0, got {total}'
+            )
+
+    def apply(self, data: 'DGData') -> Tuple['DGData', 'DGData', 'DGData']:  # type: ignore
+        return data, data, data
```
Comment (ntgbaoo) [inline tgm/split.py]: What is the behaviour of RatioSplit if the ratios passed are summed up to less than 1 (e.g 0.5/0.15/0.15)?

File: tgm/data.py [hunk 2]
```diff
@@ -482,48 +496,28 @@ def _check_pandas_import(min_version_number: str | None = None) -> None:
...
-                err_msg = 'User requires tgb to initialize a DGraph from a tgb dataset '
-                raise ImportError(err_msg)
-
-        LinkPropPredDataset, NodePropPredDataset = _check_tgb_import()
+    def from_tgb(cls, name: str, **kwargs: Any) -> DGData:
+        try:
+            from tgb.linkproppred.dataset import LinkPropPredDataset
+            from tgb.nodeproppred.dataset import NodePropPredDataset
+        except ImportError:
+            raise ImportError('TGB required to load TGB data, try `pip install py-tgb`')
```
Comment (ntgbaoo) [inline tgm/data.py:508]: Is it 'TGM required to load TGB data, try `pip install py-tgb`'?

File: tgm/split.py
Comment (ntgbaoo) [inline tgm/split.py:1]: So users need to define the implementation of `apply()` functions themselves?

File: tgm/data.py (hunk 2 above)
Comment (Jacob-Chmura) [reply tgm/data.py:508]: Im trying to say, the user needs _TGB_ in order to use TGB datasets

File: tgm/split.py
Comment (Jacob-Chmura) [reply tgm/split.py:1]: I just put in stubs for now, we will implement the apply functions before merging.

If a user wants to define their own split strategy, then yes, they'll have to implement apply.

File: tgm/split.py (hunk 1 above)
Comment (Jacob-Chmura) [reply tgm/split.py]: True, ill see what libraries like sklearn do regarding these data splits, and follow suite

Comment (Jacob-Chmura) [reply tgm/split.py]: Similarly, if test time is < val time in the timesplitter

File: tgm/data.py [hunk 3]
```diff
@@ -30,6 +31,8 @@ class DGData:
 
     static_node_feats: Tensor | None = None  # [num_nodes, D_node_static]
 
+    _split_strategy: SplitStrategy | None = None
```
Comment (Jacob-Chmura) [inline tgm/data.py:34]: Should add repr(False), equal(False)

File: tgm/data.py [hunk 4]
```diff
@@ -124,13 +127,17 @@ def _assert_tensor_is_integral(x: Tensor, name: str) -> None:
...
-            ):
+            if self.static_node_feats.ndim != 2:
                 raise ValueError(
-                    'static_node_feats must have shape [num_nodes, D_node_static], '
-                    f'got {num_nodes} nodes and shape {self.static_node_feats.shape}'
+                    f'static_node_feats must be a 2D tensor of shape [N, D_node_static], '
+                    f'but got ndim={self.static_node_feats.ndim} with shape {self.static_node_feats.shape}'
+                )
+
+            if self.static_node_feats.shape[0] < num_nodes:
```
Comment (Jacob-Chmura) [inline tgm/data.py:136]: We changed this from strict equality check to accomodate the fact that post `data.split()` the _train/val/test_ will have a static node feature reference which is a _superset_ of the nodes on that single split. 

Of course, this, as everywhere else, assumes the number of nodes is determined by the max node it.

File: tgm/data.py [hunk 5]
```diff
@@ -583,3 +577,11 @@ def _check_tgb_import() -> tuple['LinkPropPredDataset', 'NodePropPredDataset']:
             dynamic_node_feats=dynamic_node_feats,
             static_node_feats=static_node_feats,
         )
+
+        split_bounds = {
+            'train': (int(train_times.min().item()), int(train_times.max().item())),
+            'val': (int(val_times.min().item()), int(val_times.max().item())),
+            'test': (int(test_times.min().item()), int(test_times.max().item())),
+        }
```
Comment (Jacob-Chmura) [inline tgm/data.py]: It is sufficient to just store the the upper and lower bound edge times for each split. Alternatively, we could store the entire edge timestamp mask, but this would require more complex logic in the `TGBSplit` application

File: tgm/split.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            edge_start_time, edge_end_time = self.split_bounds[split_name]
+            edge_mask = (data.timestamps[data.edge_event_idx] >= edge_start_time) & (
+                data.timestamps[data.edge_event_idx] <= edge_end_time
+            )
+
+            node_mask = None
+            if data.node_ids is not None:
+                node_times = data.timestamps[data.node_event_idx]
+                if edge_mask.any():
+                    node_mask = (node_times >= (edge_start_time - 1)) & (
```
Comment (Jacob-Chmura) [inline tgm/split.py:141]: Using the weird tgb - 1 logic here akin to: https://github.com/tgm-team/tgm/blob/5bec4b724b62cd0a40c7b6ca519071edfeb54826/tgm/data.py#L531-L542

Not ideal that it's duplicated

File: examples/linkproppred/gclstm.py [hunk 7]
```diff
@@ -145,14 +145,13 @@ def eval(
 args = parser.parse_args()
 seed_everything(args.seed)
 
-train_data = DGData.from_tgb(args.dataset, split='train').discretize(args.time_gran)
-train_dg = DGraph(train_data, train_data.time_delta, device=args.device)
 
-val_data = DGData.from_tgb(args.dataset, split='val').discretize(args.time_gran)
-val_dg = DGraph(val_data, val_data.time_delta, device=args.device)
+data = DGData.from_tgb(args.dataset).discretize(args.time_gran)
```
Comment (shenyangHuang) [inline examples/linkproppred/gclstm.py]: interesting. 

if you do 
```
train_data, _ , _  = data.split()
train_data.discretize()
```

is it equivalent to 
```
data.discretize()
train_data, _ , _  = data.split()
```

Comment (shenyangHuang) [reply examples/linkproppred/gclstm.py]: should we standardize this also across examples?
